catalogs/catalog-fileset/src/main/java/org/apache/gravitino/catalog/fileset/FilesetCatalogOperations.java

```diff
@@ -1248,10 +1298,59 @@ private boolean hasCallerContext() {
         && !CallerContext.CallerContextHolder.get().context().isEmpty();
   }
 
+  @VisibleForTesting
+  FileSystem getFileSystemWithCache(Path path, Map<String, String> conf) {
+    String pathString = path.toString();
+    // extract the prefix of the path to use as the cache key
+    String prefix = extractPrefix(pathString);
+    return fileSystemCache.get(
+        new FileSystemCacheKey(conf, prefix),
+        cacheKey -> {
+          try {
+            return getFileSystem(path, conf);
+          } catch (IOException e) {
+            throw new GravitinoRuntimeException(
+                e, "Failed to get FileSystem for fileset: path: %s, conf: %s", path, conf);
+          }
+        });
+  }
+
+  /**
+   * Extracts the prefix from the given path. The prefix is defined as the scheme and the first
+   * slash after the scheme.
+   *
+   * @param path the path from which to extract the prefix.
+   * @return the prefix of the path, or an empty string if the path is null or empty.
+   */
+  @VisibleForTesting
+  String extractPrefix(String path) {
+    if (path == null || path.isEmpty()) {
+      return "";
+    }
+
+    if (path.startsWith("file:/")) {
+      return "file:///";
+    }
+
+    String protocolSlash = "://";
+    int protocolStart = path.indexOf(protocolSlash);
+    if (protocolStart == -1) {
+      return path;
+    }
+
+    int firstSlash = path.indexOf('/', protocolStart + protocolSlash.length());
+    if (firstSlash == -1) {
+      return path + "/";
+    }
+
+    return path.substring(0, firstSlash + 1);
```

Comment:
can we use `path.toUri().getAuthority()` to simplify the function logic?

Reply:
@mchades 
Please help to take a look again.

Reply:
let me think a bit.